attest/README.md

```diff
@@ -1,3 +1,108 @@
+
+The easeiest way to build an authenticated channel is to use a key exchange protocol. What this means is, Alice and Bob (from above) would take advantage of the fact the `EREPORT` structure contains 64 bytes of user-specified data. That is, when the Alice Enclave calls `EREPORT`, they provide the bytes of their public key as user-specified data. That data is then baked into the `EREPORT` structure which they get back from the CPU (and therefore, cryptographically authenticated by the CPU as coming from Alice).
+
+Similarly, Bob puts his own public key into his `EREPORT` call as user-data, and so the report he gets back contains his public key, cryptographically authenticated by the CPU. Alice and Bob then exchange the reports in the usual "local attestation" way described above.
+
+### Key Exchange
+
+At this point, Alice and Bob both know the other enclave exists, *and* (thanks to the user data) knows a public key associated with the other enclave. At this point, Alice will combine her private key with Bob's public key in order to create a new number. Bob, for his part, will combine his private key with Alic's public key to create a new number, and thanks to the math used, the "new number" they both come up with will be the same.
+
+Because both Alice and Bob are using an asymmetric cryptosystem (where you can't figure out what private key is behind a public key), and neither one ever sends their private keys, the "new number" they come up with is a "shared secret". The "Eve" passing messages back and forth doesn't have either Alice's or Bob's private keys to multiply with the public key.
+
+Bob and Alice now have a secret which each of them knows, but Eve cannot know, which means they can use this shared secret to create their own message authentication codes, between each other, without relying on special CPU instructions like `EREPORT`. When Bob wants to send a message to Alice, he can compute the authentication code for it (the MAC), and send both of them to Alice. Alice, for her part, will independently compute the MAC based on the message, and then compare her MAC with the one Bob sent over. If they match, then she knows that Eve didn't change the message.
+
+In other words, Alice can now authenticate messages from Bob, and visa-versa. If Alice and Bob also want to *hide* their messages from Eve, then Bob would first encrypt his message, then compute the MAC of the encrypted text. It's usually easier to both encrypt and MAC, because there is a special kind of encryption "Authenticated Encryption" that combines encryption and authentication codes into a single package.
```

Comment:
I think this might be better served by a separate section explaining that client-to-node communications work in basically the same way.

Reply:
This section felt confusing because enclaves on the same machine aren't opening a communication channel with each other in practice. Better to reference the most 'up-to-date' concepts in the readers' minds - AVRs and client-to-enclave.

Reply:
OK, I'll take a stab at making this more clear.